developer/src/kmc-keyboard-info/src/index.ts

```diff
@@ -0,0 +1,412 @@
+/**
+ * Merges a source .keyboard_info file with metadata extracted from .kps file and
+ * compiled files to produce a comprehensive .keyboard_info file.
+ */
+
+import { minKeymanVersion } from "./min-keyman-version.js";
+import { KeyboardInfoFile, KeyboardInfoFileIncludes, KeyboardInfoFileLanguage, KeyboardInfoFilePlatform } from "./keyboard-info-file.js";
+import { KeymanFileTypes, CompilerCallbacks, KmpJsonFile, KmxFileReader, KMX, KeymanTargets } from "@keymanapp/common-types";
+import { KeyboardInfoCompilerMessages } from "./messages.js";
+import langtags from "./imports/langtags.js";
+
+const regionNames = new Intl.DisplayNames(['en'], { type: "region" });
+const scriptNames = new Intl.DisplayNames(['en'], { type: "script" });
+const langtagsByTag = {};
+
+/**
+ * Build a dictionary of language tags from langtags.json
+ */
+
+function init(): void {
+  if(langtagsByTag['en']) {
+    // Already initialized, we can reasonably assume that 'en' will always be in
+    // langtags.json.
+    return;
+  }
+
+  for(const tag of langtags) {
+    langtagsByTag[tag.tag] = tag;
+    langtagsByTag[tag.full] = tag;
+    if(tag.tags) {
+      for(const t of tag.tags) {
+        langtagsByTag[t] = tag;
+      }
+    }
+  }
+}
+
+/* c8 ignore start */
+export class KeyboardInfoOptions {
+  /** The identifier for the keyboard */
+  keyboard_id: string;
+
+  /** The data from the .kps file, transformed to kmp.json */
+  kmpJsonData: KmpJsonFile.KmpJsonFile;
+
+  /** The path in the keymanapp/keyboards repo where this keyboard may be found (optional) */
+  sourcePath?: string;
+
+  /** The full URL to the keyboard help, starting with https://help.keyman.com/keyboard/ (optional) */
+  helpLink?: string;
+
+  /** The compiled keyboard filename and relative path (.js only) */
+  keyboardFileNameJs?: string;
+
+  /** The compiled package filename and relative path (.kmp) */
+  kmpFileName: string;
+
+  /** The source package filename and relative path (.kps) */
+  kpsFileName: string;
+};
+/* c8 ignore stop */
+
+export class KeyboardInfoCompiler {
+  constructor(private callbacks: CompilerCallbacks) {
+    init();
+  }
+
+  /**
+   * Merges source .keyboard_info file with metadata from the keyboard and package source file.
+   * This function is intended for use within the keyboards repository. While many of the
+   * parameters could be deduced from each other, they are specified here to reduce the
+   * number of places the filenames are constructed.
+   * For full documentation, see:
+   * https://help.keyman.com/developer/cloud/keyboard_info/
+   *
+   * @param sourceKeyboardInfoFileName  Path for the source .keyboard_info file
+   * @param options                     Details on files from which to extract additional metadata
+   */
+  public writeMergedKeyboardInfoFile(
+    sourceKeyboardInfoFileName: string,
+    options: KeyboardInfoOptions
```

Comment:
As I continue parsing this function for the review... man, `options` here sure does some heavy lifting.  Of particular note: `options.kmpJsonData` gets used a few times.  That's practically a main source that would normally be an explicit parameter.  I'm not against it being within `options`... but it'd definitely be helpful to acknowledge that we're relying on it existing within `options` for this method, given how related `.keyboard_info` and `kmp.json` files are.

> `Details on files from which to extract additional metadata`

> `additional metadata`

`kmp.json` metadata is hardly "additional", at least in my eyes.

Reply:
Yes, can do.